pkg/services/syncer_custom.go

```diff
@@ -0,0 +1,145 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package services
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+
+	upgradev1 "github.com/rancher/system-upgrade-controller/pkg/apis/upgrade.cattle.io/v1"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	provv1 "github.com/rancher-sandbox/rancheros-operator/pkg/apis/rancheros.cattle.io/v1"
+	"github.com/rancher-sandbox/rancheros-operator/pkg/clients"
+)
+
+type CustomSyncer struct {
+	upgradev1.ContainerSpec
+}
+
+func (j *CustomSyncer) toContainers() []corev1.Container {
+	return []corev1.Container{
+		{
+			VolumeMounts: []corev1.VolumeMount{{Name: "output",
+				MountPath: "/output",
+			}},
+			Name:    "runner",
+			Image:   j.Image,
+			Command: j.Command,
+			Args:    j.Args,
+			EnvFrom: j.EnvFrom,
+			Env:     j.Env,
+		},
+	}
+}
+
+func (j *CustomSyncer) sync(r chan interface{}, s provv1.ManagedOSVersionChannel, c *clients.Clients) ([]provv1.ManagedOSVersion, error) {
+	logrus.Infof("Syncing '%s/%s'", s.Namespace, s.Name)
+
+	serviceAccount := false
+	p, err := c.Core.Pod().Get(s.Namespace, s.Name, v1.GetOptions{})
+	if err != nil {
+		_, err = c.Core.Pod().Create(&corev1.Pod{
+			TypeMeta: v1.TypeMeta{
+				APIVersion: "v1",
+				Kind:       "Pod",
+			},
+			ObjectMeta: v1.ObjectMeta{
+				Name:      s.Name,
+				Namespace: s.Namespace,
+				OwnerReferences: []v1.OwnerReference{
+					*v1.NewControllerRef(&s, provv1.SchemeGroupVersion.WithKind("ManagedOSVersionChannel")),
+				},
+			},
+			Spec: corev1.PodSpec{
+				RestartPolicy:                corev1.RestartPolicyOnFailure,
+				AutomountServiceAccountToken: &serviceAccount,
+				InitContainers:               j.toContainers(),
+				Volumes: []corev1.Volume{{
+					Name:         "output",
+					VolumeSource: corev1.VolumeSource{EmptyDir: &corev1.EmptyDirVolumeSource{}},
+				}},
+				Containers: []corev1.Container{{
+					VolumeMounts: []corev1.VolumeMount{{Name: "output",
+						MountPath: "/output",
+					}},
+					Name:    "pause",
+					Image:   "busybox",
+					Command: []string{"/bin/sh", "-c"},
+					Args:    []string{"cat /output/data && sleep 9000000000"},
+				},
+				},
+			},
+		})
+
+		// Requeueing
+		r <- nil
+		return nil, err
+	}
+
+	// TODO Check container state before getting logs.
+	terminated := len(p.Status.InitContainerStatuses) > 0 && p.Status.InitContainerStatuses[0].Name == "runner" &&
+		p.Status.InitContainerStatuses[0].State.Terminated != nil && p.Status.InitContainerStatuses[0].State.Terminated.ExitCode == 0
+	failed := len(p.Status.InitContainerStatuses) > 0 && p.Status.InitContainerStatuses[0].Name == "runner" &&
+		p.Status.InitContainerStatuses[0].State.Terminated != nil && p.Status.InitContainerStatuses[0].State.Terminated.ExitCode != 0
+	if !terminated {
+		logrus.Infof("Waiting for '%s/%s' to finish", p.Namespace, p.Name)
+		return nil, err
```

Comment:
Feel free to ignore, I know this is a testing draft.

Shouldn't it be requed here? I don't see any logic waiting for the pod to finalize, so if this is a polling mechanism shouldn't it requeue the sync function until it can read the logs?

Reply:
good point, I missed re-enqueuing here when I've added the re-queue mechanism :sweat_smile: